dot/parachain/statement-distribution/groups.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2025 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package statementdistribution
+
+import (
+	"github.com/ChainSafe/gossamer/dot/parachain/backing"
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+)
+
+// groups tracks Validator groups within a session, plus some helpful indexing
+// for looking up groups by validator indices or authority discovery ID.
+type groups struct {
+	groups           [][]parachaintypes.ValidatorIndex
+	byValidatorIdx   map[parachaintypes.ValidatorIndex]parachaintypes.GroupIndex
+	backingThreshold uint32
+}
+
+func newGroups(initGroups [][]parachaintypes.ValidatorIndex, backingThreshold uint32) *groups {
+	if initGroups == nil {
+		initGroups = [][]parachaintypes.ValidatorIndex{}
+	}
+
+	bvi := make(map[parachaintypes.ValidatorIndex]parachaintypes.GroupIndex)
+
+	for groupIdx, group := range initGroups {
+		for _, validatorIdx := range group {
+			bvi[validatorIdx] = parachaintypes.GroupIndex(groupIdx)
+		}
+	}
+
+	return &groups{
+		groups:           initGroups,
+		byValidatorIdx:   bvi,
+		backingThreshold: backingThreshold,
+	}
+}
+
+func (g *groups) all() [][]parachaintypes.ValidatorIndex {
+	return g.groups
+}
+
+func (g *groups) group(groupIndex parachaintypes.GroupIndex) []parachaintypes.ValidatorIndex {
+	if int(groupIndex) > len(g.groups)-1 {
+		return nil
```

Comment:
is it better to return error rather than nil here? returning nil seems to be a bit ambiguous

Reply:
This mirrors the behavior of [the Rust code](https://github.com/paritytech/polkadot-sdk/blob/7cbe0c76ef8fd2aabf9f07de0156941ce3ed44b0/polkadot/node/network/statement-distribution/src/v2/groups.rs#L59), which returns an `Option<&[ValidatorIndex]>`.

Returning an error would change the semantics of the method. I assume trying to access a group that is not tracked yet is not considered an error.

However, I just noticed that I named the method `groups` instead of `get` for some reason. I'll change that.

Reply:
that makes sense.